Implement TFChainError in chain client

Mahmoud-Emad opened the pull request:
### Changes

- Created a new Error wrapper to throw errors based on their type.
- Determined the type of error and threw it accordingly.
- Updated the error interface to differentiate between Module, Token, and other error types.
- Remove the TFChain error from the types package.

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3071
- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3072

### Checklist

- [x] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

Mahmoud-Emad committed:
feat: Implement TFChainError in chain client

- Created a new Error wrapper to throw errors based on their type.
- Determined the type of error and threw it accordingly.
- Updated the error interface to differentiate between Module, Token, and other error types.

Mahmoud-Emad committed:
chore: Remove the TFChain error from thr types package.

Mahmoud-Emad committed:
chore: Replace the TFChainError of the types by the newly implemented the TFChainError class.

Mahmoud-Emad committed:
chore: Fix build

0oM4R commented on packages/tfchain_client/src/errors.ts:
```diff
@@ -0,0 +1,140 @@
+  docs?: string[];
+}
+
+class TFChainError extends Error implements ITFChainError {
```
i think this should be in types package.

0oM4R commented:
I see we are going to move the tfchain error from the types package to the tfchain package, is there a reason for that, as we discussed before we need to only refactor the class, and for the other issue we could add a method to that class to resolve the error.

Mahmoud-Emad committed:
Merge branch 'development' into development_tfchain_error

Mahmoud-Emad committed:
chore: Refactor TFChainError implementation

- Moved TFChainError into the types package.
- Updated imports to reflect the new location of TFChainError.

Mahmoud-Emad commented:
> I see we are going to move the tfchain error from the types package to the tfchain package, is there a reason for that, as we discussed before we need to only refactor the class, and for the other issue we could add a method to that class to resolve the error.



As discussed with @AhmedHanafy725 , we'll create an interface with specific fields to capture error details. We'll populate these fields based on the error type. Placing the changes in the types package makes sense yeah. Please review the changes.

0oM4R commented on packages/types/src/errors/tfchain/index.ts:
```diff
@@ -1,6 +1,155 @@
+   */
+  throw(): TFChainError {
+    const args = this.extrinsic.method.args;
+    const method = this.extrinsic.method.method;
```
Please don't use the extrinsic data, it may invalid error data https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2896, it happened before and we decided to get the exact error method https://github.com/threefoldtech/tfgrid-sdk-ts/pull/2995

we can't rely on the extrinsic

Mahmoud-Emad committed:
fix: Check on the TFChain error before throwing it from the dashboard:
- Updated the if check to check on the 'KeyError'

Mahmoud-Emad committed:
feat: Enhance the TFChainError Wrapper:

- Updated the class to get the args from the extrinsic
- Updated the class to get the metho from the extrinsic
- Updated the class to get the section from the extrinsic
- Handle the error message and added the section, method, and args into it

0oM4R commented on packages/types/src/errors/readme.md:
```diff
@@ -52,7 +52,6 @@ Also, in some cases, we have an error that comes from one of the clients, but we
 
 #### Generic Errors
 
-0. TFChainError
```
can we return the TFChain error docs please ?

0oM4R commented on packages/types/src/errors/tfchain/index.ts:
```diff
@@ -1,6 +1,163 @@
+        });
+      } else if (this.error.isToken) {
+        const tokenError = this.error.asToken.toHuman();
+        console.log({ tokenError });
```
please remove the log line

0oM4R commented on packages/types/src/errors/tfchain/index.ts:
```diff
@@ -1,6 +1,163 @@
+   *   - If the message includes "Inability to pay some fees", throws an InsufficientBalanceError.
+   *   - Otherwise, throws a TFChainError for a generic error.
+   * - If the error is not a dispatch error or an instance of Error, throws a TFChainError for an unknown error.
+   * @returns {TFChainError} The specific TFChainError thrown based on the encountered error type.
```
do we actually return an error, i think we are always throw, ?

0oM4R commented on packages/types/src/errors/tfchain/index.ts:
#### suggestion 

i notice that when throwing generic error we have something like this 
![Screenshot from 2024-08-06 12-39-25](https://github.com/user-attachments/assets/c5a9eac5-062e-4e7e-80b0-fdaaeb30f34e)

what i suggest :
1- set keyError default value to `Generic Error`
2- hide the undefined props,

we can do that by remove the class properties and if it is passed use `Object.DefineProperty`

```ts
class TFChainError extends Error implements ITFChainError {
  constructor(options: ITFChainError) {
    super(options.message);
    this.name = "TFChainError";
    options.keyError && Object.defineProperty(this, "keyError", options.keyError);
    options.section && Object.defineProperty(this, "section", options.section);
    options.args && Object.defineProperty(this, "args", options.args);
    options.method && Object.defineProperty(this, "method", options.method);
    options.docs && Object.defineProperty(this, "docs", options.docs);
  }
}
```
this will hide the undefined props:
![image](https://github.com/user-attachments/assets/85009c7d-81e8-4e33-888f-bb22c3b6f38b)

but this have <IMO> one consequenc: we won't have autcomplete and will be diffecult to access the props if it is defined, so the solution is to define an interface with the same name of the class with the optional props 
```ts
interface TFChainError {
  keyError?: string;
  section?: string;
  method?: string;
  args?: AnyTuple | string[];
  docs?: string[];
}

```
now we will be able to access them normally 
![Screenshot from 2024-08-06 13-17-46](https://github.com/user-attachments/assets/0e1bc0d8-2fe9-40cf-b3fd-3a628780fd22)

Mahmoud-Emad replied:
In my honest :), it's not good to follow, why? because we have a class that implements an interface, you have to follow the `interface`, there are no blockers with the `undefined` value, make it simple ya Omar xD :P

0oM4R replied:
Do you mean the 'ITFChainError'? 
We can remove the optional values from the interface and follow it, anyways my suggestion is for better look only.

0oM4R reviewed (commented):
thank you for the great job ya Omda

Mahmoud-Emad committed:
docs: Updated the README file with the TFChainError class

Mahmoud-Emad committed:
fix: Apply comments:

- Remove the console.log
- Remove the the undefined values from the interface

Mahmoud-Emad committed:
chore: Apply comments:

- Defined a new interface with the same name of the 'TFChainError' class to offer autcomplete
- Added defineProperty to define only the defined values

0oM4R commented on packages/types/src/errors/readme.md:
```diff
@@ -95,11 +94,26 @@ For Errors that are related to workloads and deployments operations
+
+The `TFChainErrorWrapper` class processes `DispatchError` and other errors, throwing `TFChainError` with relevant details. This wrapper class ensures that all types of errors encountered during `TFChain` operations are appropriately captured and detailed error messages are thrown
+
+#### How to use it
```
please add the case of using TFChainError 
like i this case 
```ts 
throw new TFChainError({
        message: `Error getting current block number for contract ${options.id} deletion: ${err}`,
      });

```

AhmedHanafy725 commented on packages/types/src/errors/tfchain/index.ts:
this file should be in tfchain client as it uses some types from polkadot which is not installed here and shouldn't be installed to as minimum as possible

AhmedHanafy725 replied:
please move the docs with it too.

Mahmoud-Emad committed:
chore: Apply comments:

- Updated the import of the 'TFChainError' class
- Updated the TFChain README file with some hints about 'TFChainError'

Mahmoud-Emad committed:
Merge branch 'development' into development_tfchain_error
